### PR TITLE
[core] Return source and layer ownership

### PR DESCRIPTION
**Pull request by boundsj**
When a source or layer is removed transfer ownership back to the caller so it can (optionally) take it.

ref https://github.com/mapbox/mapbox-gl-native/issues/6959

cc @jfirebaugh

### CODE REVIEW

**Comment by boundsj** [comment]
Once this is ready, I'd like to merge it to master and then cherry-pick to the iOS (and maybe Android) release branches where the returned objects can actually be used in the platform source and layer wrappers.

File: src/mbgl/map/map.cpp [hunk 1]
```diff
@@ -852,11 +852,12 @@ void Map::addSource(std::unique_ptr<style::Source> source) {
     }
 }
 
-void Map::removeSource(const std::string& sourceID) {
+std::unique_ptr<Source> Map::removeSource(const std::string& sourceID) {
     if (impl->style) {
         impl->styleMutated = true;
-        impl->style->removeSource(sourceID);
+        return impl->style->removeSource(sourceID);
```
Comment (jfirebaugh) [inline src/mbgl/map/map.cpp]: Trailing whitespace.

File: src/mbgl/map/map.cpp [hunk 2]
```diff
@@ -881,18 +882,20 @@ void Map::addLayer(std::unique_ptr<Layer> layer, const optional<std::string>& be
...
     }
 
     impl->styleMutated = true;
     impl->backend.activate();
 
-    impl->style->removeLayer(id);
+    auto removedLayer = impl->style->removeLayer(id);
     impl->onUpdate(Update::Classes);
 
     impl->backend.deactivate();
```
Comment (jfirebaugh) [inline src/mbgl/map/map.cpp]: We will need to add special logic for `CustomLayer`.  Currently, `~CustomLayer` calls the `CustomLayerDeinitializeFunction`, which may make OpenGL calls. This is the reason that we call `impl->backend.activate()/deactivate()` -- the context needs to be active when initialization/deinitialization happens.

I think we need to preserve the behavior that removing a CustomLayer triggers deinitialization, even when returning ownership. Let's do this in `Style::removeLayer`, making it parallel `Style::addLayer`, which does initialization.

Comment (boundsj) [inline src/mbgl/map/map.cpp]: 00c5210

File: src/mbgl/style/layers/custom_layer_impl.cpp [hunk 3]
```diff
@@ -43,6 +43,11 @@ void CustomLayer::Impl::initialize() {
     initializeFn(context);
 }
 
+void CustomLayer::Impl::deinitialize() {
+    assert(deinitializeFn);
+    deinitializeFn(context);
```
Comment (jfirebaugh) [inline src/mbgl/style/layers/custom_layer_impl.cpp]: Also need to remove deinitialization from `CustomLayer::Impl::~Impl`. And from the looks of that method, it's possible that `deinitializeFn` may be optional? I don't remember but we should probably keep the guard.

**Comment by boundsj** [comment]
@jfirebaugh can you please double check that https://github.com/mapbox/mapbox-gl-native/pull/7014/commits/554abe9eab82251cf5e0316040cb1fc7d1d3f26b looks ok to you?

**Comment by jfirebaugh** [comment]
👍

**Pull request merged**